File: M1/src/credit.c
#include <stdio.h>

void sum_digits(int, long, int*, int*);

int main(){

  long num;

  do{
    printf("Digit a credit card number: ");
    scanf("%ld", &num);
  } while(num < 1000000000000);

  int sum = 0;
  int first_numbers;
  
  sum_digits(2, num, &sum, &first_numbers);
  sum_digits(1, num, &sum, &first_numbers);

  if(sum % 10) puts("Invalid");
  else if(first_numbers / 10 == 4) puts("Visa");
  else if(first_numbers >= 51 && first_numbers <= 55) puts("MasterCard");
  else if(first_numbers == 34 || first_numbers == 37) puts("Amex");
  else puts("Unknow card");

  return 0;
}

void sum_digits(int first, long temp, int *sum, int *first_numbers){
  int digit;
  for(int i = first; temp > 0; i++){
    if(i % 2){
      digit = first * (temp % 10);
      if(digit >= 10) *sum += digit % 10 + digit / 10;
      else *sum += digit;
    }
    temp /= 10;
    if(temp < 100 && temp > 10) *first_numbers = temp;
  }
}

File: M6/credit.py
from std import getInt

def main():
    while True:
        value = str(getInt("Number: "))
        if len(value) >= 13 and len(value) <= 16:
            break
        else:
            print("That's not a valid card number!")
    luhn(value)

def luhn(number):
    sum = 0
    for i in range(len(number)):
        if i < len(number)//2:
            num = int(number[len(number) - 2 * (i + 1)]) * 2
        else:
            num = int(number[len(number) - (2 * (i%(len(number)//2)) + 1)])
        if num < 10:
            sum += num
        else:
            for j in range(2):
                sum += int(str(num)[j])
    if sum%10:
        print("Invalid!")
    else:
        first_numbers = int(number[0:2])
        if first_numbers >= 40 and first_numbers <= 49:
            print("VISA")
        elif first_numbers >= 34 and first_numbers <= 37:
            print("AMEX")
        elif first_numbers >= 51 and first_numbers <= 55:
            print("MASTERCARD")
        else:
            print("Invalid")

if __name__ == "__main__":
    main()
